core/components/fixed_block.hpp

```diff
@@ -0,0 +1,217 @@
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_COMPONENTS_FIXED_BLOCK_HPP_
+#define GKO_CORE_COMPONENTS_FIXED_BLOCK_HPP_
+
+
+#include <algorithm>
+#include <type_traits>
+#include <vector>
+
+
+#include <ginkgo/core/base/exception.hpp>
+#include <ginkgo/core/base/math.hpp>
+
+
+namespace gko {
+namespace blockutils {
+
+
+/**
+ * @brief A dense block of values with compile-time constant dimensions
+ *
+ * The blocks are interpreted as row-major. However, in the future,
+ *  a layout template parameter can be added if needed.
+ *
+ * The primary use is to reinterpret subsets of entries in a big array as
+ *  small dense blocks.
+ *
+ * @tparam ValueType  The numeric type of entries of the block
+ * @tparam nrows  Number of rows
+ * @tparam ncols  Number of columns
+ */
+template <typename ValueType, int nrows, int ncols>
+class FixedBlock final {
+    static_assert(nrows > 0, "Requires positive number of rows!");
+    static_assert(ncols > 0, "Requires positive number of columns!");
+
+public:
+    using value_type = ValueType;
+
+    value_type &at(const int row, const int col)
+    {
+        return vals[row * ncols + col];
+    }
+
+    const value_type &at(const int row, const int col) const
+    {
+        return vals[row * ncols + col];
+    }
+
+    value_type &operator()(const int row, const int col)
+    {
+        return at(row, col);
+    }
+
+    const value_type &operator()(const int row, const int col) const
+    {
+        return at(row, col);
+    }
+
+private:
+    ValueType vals[nrows * ncols];
+};
+
+
+/**
+ * A lightweight dynamic block type on the host
+ *
+ * @tparam ValueType The numeric type of entries of the block
+ */
+template <typename ValueType>
+class DenseBlock final {
+public:
+    using value_type = ValueType;
+
+    /**
+     * If this default construtor is used, @ref set_executor must be called
+     *  prior to any other member function.
+     */
+    DenseBlock() : nrows_{0}, ncols_{0} {}
+
+    DenseBlock(const int num_rows, const int num_cols)
+        : nrows_{num_rows}, ncols_{num_cols}, vals_(num_rows * num_cols)
+    {}
+
+    value_type &at(const int row, const int col)
+    {
+        return vals_[row * ncols_ + col];
+    }
+
+    const value_type &at(const int row, const int col) const
+    {
+        return vals_[row * ncols_ + col];
+    }
+
+    value_type &operator()(const int row, const int col)
+    {
+        return at(row, col);
+    }
+
+    const value_type &operator()(const int row, const int col) const
+    {
+        return at(row, col);
+    }
+
+    int size() const { return nrows_ * ncols_; }
+
+    void resize(const int nrows, const int ncols)
+    {
+        vals_.resize(nrows * ncols);
+        nrows_ = nrows;
+        ncols_ = ncols;
+    }
+
+    void zero()
+    {
+        std::fill(vals_.begin(), vals_.end(), static_cast<value_type>(0));
+    }
+
+private:
+    int nrows_;
+    int ncols_;
+    std::vector<value_type> vals_;
```

Comment:
I tried to use `Array`. The first issue with that is it needs extra code to pass in the executor. Even after adding code for that, I couldn't get it to work nicely in the core code. I use it in `Fbcsr::read`. This will only ever be used on the host, and I think using `std::vector` addresses the safely concerns raised by @tcojean .

Reply:
I see. Maybe then you can put it in the `detail` namespace and add an `@internal` tag with a note that says that this is only used in the read function ? You already add a `final` qualification, so that is good.